### PR TITLE
[Composable] Add hybrid shard AC compile test

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105207
* #105206
* #105205
* #105180
* #105176

This was request to ensure hybrid shard + AC + compile works.

Differential Revision: [D47462393](https://our.internmc.facebook.com/intern/diff/D47462393/)

### CODE REVIEW

File: test/distributed/_composable/fully_shard/test_fully_shard_compile.py [hunk 1]
```diff
@@ -39,9 +39,11 @@ def test_compile(self):
                     ShardingStrategy.FULL_SHARD,
                     ShardingStrategy.SHARD_GRAD_OP,
                     ShardingStrategy.NO_SHARD,
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
                 ],
                 "skip_fsdp_guards": [True, False],
-                "act_checkpoint": [True],
+                "act_checkpoint": [True, False],
```
Comment (awgu) [inline test/distributed/_composable/fully_shard/test_fully_shard_compile.py:46]: As commented on in the previous PR, I think this change should be done in that PR.

File: test/distributed/_composable/fully_shard/test_fully_shard_compile.py [hunk 2]
```diff
@@ -39,9 +39,11 @@ def test_compile(self):
                     ShardingStrategy.FULL_SHARD,
                     ShardingStrategy.SHARD_GRAD_OP,
                     ShardingStrategy.NO_SHARD,
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
```
Comment (awgu) [inline test/distributed/_composable/fully_shard/test_fully_shard_compile.py:43]: How do these get tested in CI or even with a single node locally? What process groups get constructed?

Is the replication group a singleton process group and the sharding process group takes all GPUs?

**Review by awgu — APPROVED** [review]

File: test/distributed/_composable/fully_shard/test_fully_shard_compile.py (hunk 2 above)
Comment (awgu) [reply test/distributed/_composable/fully_shard/test_fully_shard_compile.py:43]: (I think the answer to this affects whether we should land this PR as is, or if we need to do some changes on manually specifying PGs.)

Comment (rohan-varma) [reply test/distributed/_composable/fully_shard/test_fully_shard_compile.py:43]: that's a good point, I will check this. I ran it on a single node so I wonder if it just created an FSDP group and singel rank data parallel groups.

Comment (rohan-varma) [reply test/distributed/_composable/fully_shard/test_fully_shard_compile.py:43]: @awgu I think we just need a similar world size change as in the prev diff.

**Review by awgu — APPROVED** [review]

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**